src/reducers/newThought.js

```diff
@@ -102,46 +109,46 @@ export default ({ at, insertNewSubthought, insertBefore, value = '', offset, pre
-    addAsContext: (showContextsParent && !insertNewSubthought) || (showContexts && insertNewSubthought),
-    rank: newRank,
-    value
-  })
-
-  if (!preventSetCursor) {
-    dispatch({
-      type: 'setCursor',
-      editing: true,
-      thoughtsRanked: (insertNewSubthought ? unroot(path) : contextOf(path)).concat({ value, rank: newRank }),
-      offset: offset != null ? offset : value.length,
-    })
-  }
+  const reducers = [
 
-  // tutorial step 1
-  if (tutorialStepNewThoughtCompleted) {
-    clearTimeout(globals.newSubthoughtModalTimeout)
-    dispatch({ type: 'tutorialNext' })
-  }
-  // some hints are rolled back when a new thought is created
-  else if (tutorialStep === TUTORIAL2_STEP_CONTEXT1_PARENT_HINT) {
-    dispatch({ type: 'tutorialStep', value: TUTORIAL2_STEP_CONTEXT1_PARENT })
-  }
-  else if (tutorialStep === TUTORIAL2_STEP_CONTEXT1_HINT) {
-    dispatch({ type: 'tutorialStep', value: TUTORIAL2_STEP_CONTEXT1 })
-  }
-  else if (tutorialStep === TUTORIAL2_STEP_CONTEXT2_PARENT_HINT) {
-    dispatch({ type: 'tutorialStep', value: TUTORIAL2_STEP_CONTEXT2_PARENT })
-  }
-  else if (tutorialStep === TUTORIAL2_STEP_CONTEXT2_HINT) {
-    dispatch({ type: 'tutorialStep', value: TUTORIAL2_STEP_CONTEXT2 })
-  }
+    // newThoughtSubmit
+    state => newThoughtSubmit(state, {
+      context: insertNewSubthought
+        ? pathToContext(thoughtsRanked)
+        : context,
+      // inserting a new child into a context functions the same as in the normal thought view
+      addAsContext: (showContextsParent && !insertNewSubthought) || (showContexts && insertNewSubthought),
+      rank: newRank,
+      value
+    }),
 
-  return {
-    rank: newRank
-  }
+    // setCursor
+    !preventSetCursor
+      ? state => setCursor(state, {
+        editing: true,
+        thoughtsRanked: (insertNewSubthought ? unroot(path) : contextOf(path)).concat({ value, rank: newRank }),
+        offset: offset != null ? offset : value.length,
+      })
+      : null,
+
+    // tutorial step 1
+    tutorialStepNewThoughtCompleted ? clearTimeout(globals.newSubthoughtModalTimeout) || (state => tutorialNext(state))
+    // some hints are rolled back when a new thought is created
+    : tutorialStep === TUTORIAL2_STEP_CONTEXT1_PARENT_HINT
+      ? state => tutorialStepReducer(state, { value: TUTORIAL2_STEP_CONTEXT1_PARENT })
+      : tutorialStep === TUTORIAL2_STEP_CONTEXT1_HINT ?
+        state => tutorialStepReducer(state, { value: TUTORIAL2_STEP_CONTEXT1 })
+        : tutorialStep === TUTORIAL2_STEP_CONTEXT2_PARENT_HINT ?
+          state => tutorialStepReducer(state, { value: TUTORIAL2_STEP_CONTEXT2_PARENT })
+          : tutorialStep === TUTORIAL2_STEP_CONTEXT2_HINT ?
+            state => tutorialStepReducer(state, { value: TUTORIAL2_STEP_CONTEXT2 })
+            : null,
+
+    // return new rank in case composed reducers need it
+    state => ({
+      newRank
+    })
```

Comment:
Won't this be adding `newRank` to the store state too ?

Reply:
Yes, it is a subpar implementation. Hence the subtask:

>  Use a selector to get rank of new thought. Adding `newRank` to the state just to pass it to the next reducer is subpar.